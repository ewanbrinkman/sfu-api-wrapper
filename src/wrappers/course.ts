import { requestSFUAcademicCalendarApiCourses } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawApiCourse,
} from '@api-types';
import { Course } from '@api';

export default async function course(
    department: string,
    number: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Course> {
    const rawApiCourse: RawApiCourse = (
        await requestSFUAcademicCalendarApiCourses(
            year,
            term,
            department,
            number,
        )
    ).data;

    return Course.fromRawApiCourse(rawApiCourse);
}
