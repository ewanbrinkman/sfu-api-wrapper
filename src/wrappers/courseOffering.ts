import { requestSFUCourseOutlinesApi } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawApiCourseOffering,
} from '@api-types';
import { CourseOffering } from '@api';

export default async function courseOutline(
    department: string,
    number: number,
    section: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<CourseOffering> {
    const rawApiCourseOffering: RawApiCourseOffering = (
        await requestSFUCourseOutlinesApi(
            year,
            term,
            department,
            number,
            section,
        )
    ).data;

    return CourseOffering.fromRawApiCourseOffering(rawApiCourseOffering);
}
