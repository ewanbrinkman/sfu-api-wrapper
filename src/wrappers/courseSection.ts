import { requestSFUCourseOutlinesApi } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawCourseSectionData,
} from '@api-types';
import { CourseSection } from '@api';

export default async function courseSection(
    department: string,
    number: string,
    section: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<CourseSection> {
    const response = await requestSFUCourseOutlinesApi(
        year,
        term.toLowerCase(),
        department.toLowerCase(),
        number.toLowerCase(),
        section.toLowerCase(),
    );
    const rawCourseSectionData: RawCourseSectionData = await response.json();

    return CourseSection.fromRawCourseSectionData(
        rawCourseSectionData,
        year,
        term,
        department,
    );
}
