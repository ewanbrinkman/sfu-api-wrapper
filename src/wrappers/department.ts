import { requestSFUAcademicCalendarApiCourses, limitConcurrency } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawDepartmentCourse,
} from '@api-types';

export default async function department(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<string[]> {
    const rawDepartmentCourses: RawDepartmentCourse[] = (
        await requestSFUAcademicCalendarApiCourses(year, term, department)
    ).data;

    return rawDepartmentCourses.map(rawDepartmentCourse => rawDepartmentCourse.value);
}
