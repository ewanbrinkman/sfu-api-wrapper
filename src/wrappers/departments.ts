import { requestSFUAcademicCalendarApiCourses } from '@utils';
import { CourseOutlinesTerm, CourseOutlinesYear } from '@api-types';

type DepartmentValue = {
    title: string;
    value: string;
};

type DepartmentValues = DepartmentValue[];

export default async function departments(
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<string[]> {
    const response = (
        await requestSFUAcademicCalendarApiCourses(year, term)
    );
    const rawDepartments: DepartmentValues = await response.json();

    return rawDepartments.map(rawDepartment => rawDepartment.value);
}
