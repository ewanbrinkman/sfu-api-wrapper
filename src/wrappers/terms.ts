import { requestSFUCourseOutlinesApi } from '@utils';
import { CourseOutlinesYear, Term } from '@api-types';

type TermValue = {
    value: Term;
};

type TermValues = TermValue[];

export default async function terms(
    year: CourseOutlinesYear = 'current',
): Promise<Term[]> {
    const response = await requestSFUCourseOutlinesApi(year);
    const rawTerms: TermValues = await response.json();

    return rawTerms.map((rawTerm) => rawTerm.value);
}
