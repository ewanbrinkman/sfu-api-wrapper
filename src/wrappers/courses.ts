import { requestSFUAcademicCalendarApiCourses, limitConcurrency } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawDepartmentCourse,
} from '@api-types';
import { Course } from '@api';
import wrappers from '@wrappers';

export default async function courses(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Course[]> {
    const response = (
        await requestSFUAcademicCalendarApiCourses(year, term, department)
    );
    const rawDepartmentCourses: RawDepartmentCourse[] = await response.json();

    // Can't request all of them at once, since that spams the API and the API
    // has problems with that.
    return await limitConcurrency(
        rawDepartmentCourses,
        async (rawDepartmentCourse) => {
            return wrappers.course(department, rawDepartmentCourse.value, year, term);
        }
    );
}
