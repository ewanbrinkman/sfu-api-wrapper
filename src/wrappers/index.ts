import course from './course';
import courseSection from './courseSection';
import departmentCourseNumbers from './departmentCourseNumbers';
import departmentCourses from './departmentCourses';
import departmentNames from './departments';
import departments from './departments';
import years from './years';
import terms from './terms';

export default {
    course: course,
    courseSection: courseSection,
    departmentCourseNumbers: departmentCourseNumbers,
    departmentCourses: departmentCourses,
    departmentNames: departmentNames,
    departments: departments,
    terms: terms,
    years: years,
};
