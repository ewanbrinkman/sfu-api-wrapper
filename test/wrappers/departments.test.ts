import wrappers from '../../src/wrappers';

describe('departments', () => {
    test('request departments', async () => {
        const departments = await wrappers.departments(2022, 'fall');
        expect(departments).toMatchSnapshot();
    }, 30000);
});
