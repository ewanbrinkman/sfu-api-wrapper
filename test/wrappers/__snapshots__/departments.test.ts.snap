// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`departments request departments 1`] = `
[
  Department {
    "id": "acma",
    "name": "Actuarial Mathematics",
  },
  Department {
    "id": "als",
    "name": "Applied Legal Studies",
  },
  Department {
    "id": "apma",
    "name": "Applied Mathematics",
  },
  Department {
    "id": "arab",
    "name": "Arabic",
  },
  Department {
    "id": "arch",
    "name": "Archaeology",
  },
  Department {
    "id": "bisc",
    "name": "Biological Sciences",
  },
  Department {
    "id": "bpk",
    "name": "Biomedical Physiology and Kinesiology",
  },
  Department {
    "id": "bus",
    "name": "Business Administration",
  },
  Department {
    "id": "ca",
    "name": "Contemporary Arts",
  },
  Department {
    "id": "chem",
    "name": "Chemistry",
  },
  Department {
    "id": "chin",
    "name": "Chinese",
  },
  Department {
    "id": "cmns",
    "name": "Communication",
  },
  Department {
    "id": "cmpt",
    "name": "Computing Science",
  },
  Department {
    "id": "cogs",
    "name": "Cognitive Science",
  },
  Department {
    "id": "crim",
    "name": "Criminology",
  },
  Department {
    "id": "data",
    "name": "Data Science",
  },
  Department {
    "id": "dial",
    "name": "Dialogue",
  },
  Department {
    "id": "dmed",
    "name": "Digital Media",
  },
  Department {
    "id": "easc",
    "name": "Earth Sciences",
  },
  Department {
    "id": "eco",
    "name": "Ecological Restoration",
  },
  Department {
    "id": "econ",
    "name": "Economics",
  },
  Department {
    "id": "edpr",
    "name": "Education Professional",
  },
  Department {
    "id": "educ",
    "name": "Education",
  },
  Department {
    "id": "engl",
    "name": "English",
  },
  Department {
    "id": "ensc",
    "name": "Engineering Science",
  },
  Department {
    "id": "env",
    "name": "Environment",
  },
  Department {
    "id": "etec",
    "name": "Educational Technology and Learning",
  },
  Department {
    "id": "evsc",
    "name": "Environmental Science",
  },
  Department {
    "id": "fal",
    "name": "Foundations of Academic Literacy",
  },
  Department {
    "id": "fan",
    "name": "Foundations of Analytical and Quantitative Reasoning",
  },
  Department {
    "id": "fass",
    "name": "Faculty of Arts and Social Sciences",
  },
  Department {
    "id": "fren",
    "name": "French",
  },
  Department {
    "id": "ga",
    "name": "Global Asia",
  },
  Department {
    "id": "geog",
    "name": "Geography",
  },
  Department {
    "id": "germ",
    "name": "German",
  },
  Department {
    "id": "gero",
    "name": "Gerontology",
  },
  Department {
    "id": "grk",
    "name": "Greek",
  },
  Department {
    "id": "gs",
    "name": "General Studies",
  },
  Department {
    "id": "gsws",
    "name": "Gender, Sexuality, and Women's Studies",
  },
  Department {
    "id": "hist",
    "name": "History",
  },
  Department {
    "id": "hsci",
    "name": "Health Sciences",
  },
  Department {
    "id": "hum",
    "name": "Global Humanities",
  },
  Department {
    "id": "iat",
    "name": "Interactive Arts and Technology",
  },
  Department {
    "id": "indg",
    "name": "Indigenous Studies",
  },
  Department {
    "id": "inlg",
    "name": "Indigenous Languages",
  },
  Department {
    "id": "ins",
    "name": "Individualized Interdisciplinary Studies",
  },
  Department {
    "id": "is",
    "name": "International Studies",
  },
  Department {
    "id": "ital",
    "name": "Italian",
  },
  Department {
    "id": "japn",
    "name": "Japanese",
  },
  Department {
    "id": "lang",
    "name": "Language",
  },
  Department {
    "id": "las",
    "name": "Latin American Studies",
  },
  Department {
    "id": "lbrl",
    "name": "Liberal Arts",
  },
  Department {
    "id": "lbst",
    "name": "Labour Studies",
  },
  Department {
    "id": "ling",
    "name": "Linguistics",
  },
  Department {
    "id": "ls",
    "name": "Liberal Studies",
  },
  Department {
    "id": "macm",
    "name": "Mathematics and Computing Science",
  },
  Department {
    "id": "masc",
    "name": "Marine Science",
  },
  Department {
    "id": "math",
    "name": "Mathematics",
  },
  Department {
    "id": "mbb",
    "name": "Molecular Biology and Biochemistry",
  },
  Department {
    "id": "mse",
    "name": "Mechatronic Systems Engineering",
  },
  Department {
    "id": "mtec",
    "name": "Management and Technology",
  },
  Department {
    "id": "neur",
    "name": "Neuroscience",
  },
  Department {
    "id": "nusc",
    "name": "Nuclear Science",
  },
  Department {
    "id": "onc",
    "name": "Oncology",
  },
  Department {
    "id": "pers",
    "name": "Persian",
  },
  Department {
    "id": "phil",
    "name": "Philosophy",
  },
  Department {
    "id": "phys",
    "name": "Physics",
  },
  Department {
    "id": "plan",
    "name": "Planning",
  },
  Department {
    "id": "plcy",
    "name": "Public Policy",
  },
  Department {
    "id": "pol",
    "name": "Political Science",
  },
  Department {
    "id": "psyc",
    "name": "Psychology",
  },
  Department {
    "id": "pub",
    "name": "Publishing",
  },
  Department {
    "id": "punj",
    "name": "Punjabi",
  },
  Department {
    "id": "rem",
    "name": "Resource and Environmental Management",
  },
  Department {
    "id": "sa",
    "name": "Sociology and Anthropology",
  },
  Department {
    "id": "sci",
    "name": "Science",
  },
  Department {
    "id": "sd",
    "name": "Sustainable Development",
  },
  Department {
    "id": "sda",
    "name": "Social Data Analytics",
  },
  Department {
    "id": "see",
    "name": "Sustainable Energy Engineering",
  },
  Department {
    "id": "span",
    "name": "Spanish",
  },
  Department {
    "id": "stat",
    "name": "Statistics",
  },
  Department {
    "id": "tekx",
    "name": "Technology - FAS",
  },
  Department {
    "id": "trss",
    "name": "Terrorism, Risk and Security Studies",
  },
  Department {
    "id": "urb",
    "name": "Urban Studies",
  },
  Department {
    "id": "wl",
    "name": "World Literature",
  },
]
`;
